utils/pdftoplotter.1
.TH pdftoplotter 1
.SH NAME
pdftoplotter \- Portable Document Format (PDF) to HPGL/SVG simple strokes
.SH SYNOPSIS
.B pdftoplotter
[options]
\-i
.IR PDF-file
\-o
.IR output-file.[hpgl | svg]
.SH DESCRIPTION
.B pdftoplotter
converts Portable Document Format (PDF) files into a series of pen strokes using any of the following output formats:
.IP \(bu
Hewlett-Packard Graphics Language (HPGL)
.IP \(bu
Scalable Vector Graphics (SVG)
.PP
.B pdftoplotter
reads the PDF file,
.IR PDF-file ,
and writes to
.IR output-file .
The
.IR -n
flag specifies which page from the PDF to render to the output file (default = 1).

.PP
If Hershey fonts are installed in /usr/share/hershey-fonts/,
.B pdftoplotter
replaces outline font character fills in the PDF with stroked
Hershey fonts.
.PP
.PP
.SH OPTIONS
.TP
.BI \-i
or
.BI \-\-input
Specifies the input filename. Required.
.TP
.BI \-o
or
.BI \-\-output
Specifies the output filename. Required.
.TP
.BI \-\-format
One of
.IR
hpgl
or
.IR
svg
. If not supplied, format taken from output file extension.
.TP
.BI \-\-verboseFonts
Prints diagnostics explaining how Hershey fonts are selected to
substitute for PDF document fonts. (Someday there will be a
configuration file that allows a user-tuned replacement policy.)
.SH ENVIRONMENT VARIABLES
.TP
.BI HERSHEY_FONT_PATH
Path to the location of the installed Hershey .jhf files.
.SH EXIT CODES
The poppler tools use the following exit codes:
.TP
0
No error.
.TP
255
Error.
.SH AUTHOR
The pdftoplotter software and documentation are copyright 1996-2004 Glyph
& Cog, LLC and copyright 2005-2020 The Poppler Developers.
.SH "SEE ALSO"
.BR pdfdetach (1),
.BR pdffonts (1),
.BR pdfimages (1),
.BR pdfinfo (1),
.BR pdftohtml (1),
.BR pdftoppm (1),
.BR pdftops (1),
.BR pdftotext (1)
.BR pdfseparate (1),
.BR pdfsig (1),
.BR pdfunite (1)


NEWS
Release 20.10.0:
        core:
         * Filter out repeated forms
         * Implement EmbedStream::reset()
         * CairoOutputDev: evict just font faces owned solely by cache.
         * Splash: Rename Yd to Ydown, Xu to Xup, etc.
         * Splash: fix crash in out-of-memory situation.
         * Fix some undefined behaviour situation with forged files

Release 20.09.0:
        core:
         * Compability fix for Forms
         * Fix fetching of Objects failing in some cases
         * Fix clearing date in Annot setModified/setDate
         * TextSelectionPainter: support glyphless fonts
         * Splash: Don't try read past end of image
         * avoid abort() on large memory allocation
         * Fix memory leak on broken files
         * Fix potential invalid memory read
         * Small code improvements

        qt5:
         * Document TextAnnotation::inplaceAlign
         * Make Annotation::setModification/CreationDate work on existing annots
         * Be a bit more stubborn converting dates that come from xml
         * Clean as many null characters from the end as possible when converting strings

        glib:
         * Add accessor functions for PopplerAttachment
         * Deprecate PopplerAttachment GTime fields
         * Deprecate PopplerDocument date properties
         * Add PopplerAnnotInk and related methods
         * Support ink annotations in the glib demo

        utils:
         * pdftoppm: report error and exit if output file cannot be written
         * Document that PDF-file can be '-' to read it from stdin

        build system:
         * cmake: Modern way to link against libpng, zlib and libtiff
         * cmake: Remove stray support for lcms1 in pdftocairo

Release 20.08.0:
        core:
         * Sub-page objects: initialize clip max values considering the render resolution. Issue #937
         * Splash: Set initial line width to 1. Issue #674
         * Fix stack overflow with specially crafted files
         * GfxShading: Simplify holding the Function
         * Splash: Fix x86 + windows asm build

        qt5:
         * Deprecate Document::toc
         * Deprecate AnnotationUtils

Release 0.90.1:
        core:
         * Fix regression on PS conversion.
           Regression only happened on applications that are locale enabled
           i.e. Okular but not pdftops, when using a quite new lcms and
           the user locale uses , as decimal separator instead of .
         * Add UTF16LE support to TextStringToUCS4.
           Even if the standard clearly says it should be UTF16BE

        qt6:
         * Add work in progress qt6 port. Ignore for now :)

Release 0.90.0:
        core:
         * Color profile tweaks
         * Small signature improvements
         * BBoxOutputDev: Fix calculation when type3 fonts are involved
         * Fix potential crash when reading Forms
         * Fix infinite loop in broken file

        glib:
         * Fix adding annots in rotated pages
         * Add ability to reset forms
         * Several fixes to the documentation

        qt5:
         * Make it clear we require Qt 5.5
         * demo: Fix crash on broken files
         * Small documentation improvements

        utils:
         * pdftoppm: Add option to set display profile
         * pdftops: Add a -rasterize option with values always, never, or whenneeded

        build system:
         * Require cmake 3.5
         * More modern cmake way to link against curl

Release 0.89.0:
        core:
         * Add support for ResetForm action. Issue #225
         * Fix crash in PDFDoc::getSignatureFields when there's no Forms at all
         * Fix exporting to PS of some files with CID fonts
         * Use ICC profiles in PS output (if new enough lcms is used)
         * Allow almost-singular tiling pattern matrices. Issue #894
         * Fix memory leak when failing to load some fonts
         * CairoOutputDev: Use stroke opacity when clipping to a stroke path
         * CairoOutputDev: Fix tiling patterns when pattern cell is too far. Issue #190

        glib:
         * Add poppler_movie_get_aspect

        cpp:
         * Add the font infos to the text_box object

Release 0.88.0:
        core:
         * Support Widget Annotation Buttons not linked to any Form
         * SplashOuputDev: Use stroking opacity when clipping to a stroke path
         * Handle 1 bit RGB images in ICC colorspace
         * Internal code improvements

        qt5:
         * Add Document::signatures. Returns signatures not attached to any page
         * ArthurOutputDev: Fix font hinting
         * ArthurOutputDev: Set the opacity when filling with axial gradients
         * ArthurOutputDev: Implement the clipToStrokePath method
         * ArthurOutputDev: Use stroking opacity when clipping to a stroke path

        glib:
         * Add poppler_page_get_bounding_box
         * Add poppler_form_field_get_alternate_ui_name
         * Implement rotation for 'flagNoRotate' annots. Issue #767

        cpp:
         * Add non_raw_non_physical layout for page::text()

        utils:
         * pdftohtml: Fix noRoundedCoordinates->noroundcoord in man page
         * pdfsig: Show also signatures that aren't attached to any page

Release 0.87.0:
        core:
         * Fix leak in broken files
         * Internal code improvements

        qt5:
         * Add option to get form choice for export value
         * ArthurOutputDev: Avoid division by zero in updateLineDash. Issue #695

        glib:
         * Internal code improvements

         utils:
         * pdftohtml: Fix memory leak in HtmlOutputDev::getLinkDest

Release 0.86.1:
        core:
         * Fix regression in Browse Link handling
         * Internal code improvements

Release 0.86.0:
        core:
         * Fix link content exfiltration attack
         * Splash: Implement gouraudTriangleShadedFill for some non parametrized shadings. Issue #881

utils/pdftoplotter.cc
//========================================================================
//
// pdftoplotter.cc
//
// Render PDF using only simple straight line-segment strokes, suitable for
// drawing on a pen plotter (or other vector device). Fonts can be substituted
// with Hershey (stroke) fonts rather than drawn as outlines.
//
// This file is licensed under the GPLv2 or later
//
// Copyright 2020 Jon Howell <jonh@jonh.net>
//
//========================================================================

#include "config.h"
#include <poppler-config.h>
#include "parseargs.h"
#include "PDFDocFactory.h"
#include "splash/SplashBitmap.h"

#include "StrokeOutputDev.h"

using namespace std;

struct Args
{
    bool help = false;
    char inputFilename[1024] = { '\0' };
    char outputFilename[1024] = { '\0' };
    char format[1024] = { '\0' };
    int pageNum = 1;
    int alarm = 0;
    bool verboseFonts = false;
} args;

static const ArgDesc argDesc[] = { { "-h", argFlag, &args.help, 0, "Display this help message" },
                                   { "-i", argString, &args.inputFilename, sizeof(args.inputFilename), "Input file name" },
                                   { "--input", argString, &args.inputFilename, sizeof(args.inputFilename), "Input file name" },
                                   { "-o", argString, &args.outputFilename, sizeof(args.outputFilename), "Output file name" },
                                   { "--output", argString, &args.outputFilename, sizeof(args.outputFilename), "Output file name" },
                                   { "--format", argString, &args.format, sizeof(args.format), "Format: hpgl | svg (default: guess from extn)" },
                                   { "-n", argInt, &args.pageNum, sizeof(args.pageNum), "Page number to render" },
                                   { "--verboseFonts", argFlag, &args.verboseFonts, false, "Print font substitution decisions." },
                                   { "--debugAlarm", argInt, &args.alarm, sizeof(args.alarm), "(debug) Stop rendering after n objects" },
                                   {} };

static int processPage()
{
    GooString *fileName = new GooString(args.inputFilename);
    PDFDoc *document = PDFDocFactory().createPDFDoc(*fileName, nullptr, nullptr);
    if (document == nullptr || !document->isOk()) {
        return -1;
    }
    if (!document->getPage(args.pageNum)) {
        cerr << "Document has no page " << args.pageNum << endl;
        return -1;
    }

    StrokeOutputDev *strokeOut = new StrokeOutputDev(args.verboseFonts);
    strokeOut->setDebugAlarm(args.alarm);
    strokeOut->startDoc(document);

    int x_resolution = 150;
    int y_resolution = 150;
    document->displayPageSlice(strokeOut, args.pageNum, x_resolution, y_resolution,
                               /*rotate*/ 0,
                               /*useMediaBox*/ true,
                               /*crop*/ true,
                               /*printing*/ false, -1, -1, -1, -1,
                               /*abortCheckCbk*/ nullptr,
                               /*annotDisplayDecideCbk*/ nullptr,
                               /*annotDisplayDecideCbkData*/ nullptr);

    SplashBitmap *bitmap = strokeOut->getBitmap();
    bitmap->writeImgFile(splashFormatPng, "proof.png", x_resolution, y_resolution);

    if (strcmp(args.format, "hpgl") == 0) {
        strokeOut->writePlotHpgl(args.outputFilename);
    }
    if (strcmp(args.format, "svg") == 0) {
        strokeOut->writePlotSvg(args.outputFilename);
    }

    delete strokeOut;
    delete document;
    delete fileName;
    return 0;
}

string getExtension(const string &s);
string getExtension(const string &s)
{
    size_t pos = s.rfind('.');
    if (pos != string::npos) {
        return s.substr(pos + 1, s.length() - pos);
    }
    return "";
}

int main(int argc, char *argv[])
{
    bool ok = parseArgs(argDesc, &argc, argv);
    if (!ok || args.help) {
        printUsage(argv[0], "", argDesc);
        return ok ? 0 : -1;
    }
    if (strlen(args.format) == 0) {
        string extn = getExtension(args.outputFilename);
        if (extn == "hpgl") {
            strcpy(args.format, "hpgl");
        } else if (extn == "svg") {
            strcpy(args.format, "svg");
        } else {
            cerr << "Cannot guess output format from extension." << endl;
            printUsage(argv[0], "", argDesc);
            return -1;
        }
    }

    globalParams = std::make_unique<GlobalParams>();
    int rc = processPage();
    return rc;
}
